File: src/translator/translator.py
"""A tiny, deterministic translator with an in-memory phrase dictionary.

This is a simple offline translator intended for tests and demonstration.
It supports 'en' <-> 'tr' for a small set of phrases and is case-insensitive.
"""
from __future__ import annotations

import logging
from typing import Dict

logger = logging.getLogger(__name__)

_TRANSLATIONS: Dict[str, Dict[str, str]] = {
    "en->tr": {
        "hello": "merhaba",
        "goodbye": "güle güle",
        "thank you": "teşekkürler",
        "yes": "evet",
        "no": "hayır",
    },
    "tr->en": {
        "merhaba": "hello",
        "güle güle": "goodbye",
        "teşekkürler": "thank you",
        "evet": "yes",
        "hayır": "no",
    },
}


class Translator:
    """Simple translator supporting a small offline dictionary.

    Usage:
        t = Translator()
        t.translate("hello", src="en", dest="tr")  # -> "merhaba"
    """

    SUPPORTED = {"en", "tr"}

    def translate(self, text: str, src: str = "en", dest: str = "tr") -> str:
        """Translate `text` from `src` to `dest`.

        - Raises ValueError if `src` or `dest` is unsupported or equal.
        - If an exact phrase match isn't found, returns the original text and logs a warning.
        """
        src = src.lower()
        dest = dest.lower()

        if src not in self.SUPPORTED:
            logger.error("Unsupported source language: %s", src)
            raise ValueError(f"Unsupported source language: {src}")
        if dest not in self.SUPPORTED:
            logger.error("Unsupported destination language: %s", dest)
            raise ValueError(f"Unsupported destination language: {dest}")
        if src == dest:
            logger.error("Source and destination language are the same: %s", src)
            raise ValueError("Source and destination must differ")

        key = f"{src}->{dest}"
        phrase = text.strip().lower()
        translation = _TRANSLATIONS.get(key, {}).get(phrase)
        if translation is None:
            logger.warning("No translation found for '%s' (%s->%s)", text, src, dest)
            return text
        return translation


# convenience function
_translator = Translator()


def translate(text: str, src: str = "en", dest: str = "tr") -> str:
    return _translator.translate(text, src=src, dest=dest)

File: src/PlayLingo/Translator.cs
using System;
using System.Collections.Generic;

namespace PlayLingo;

/// <summary>
/// Simple translator with a small, in-memory phrase dictionary.
/// </summary>
public class Translator
{
    private readonly Dictionary<string, Dictionary<string, string>> _translations;

    /// <summary>
    /// Initialize the built-in translation dictionary.
    /// </summary>
    public Translator()
    {
        _translations = new()
        {
            ["en->tr"] = new()
            {
                ["hello"] = "merhaba",
                ["goodbye"] = "güle güle",
                ["thank you"] = "teşekkürler",
                ["yes"] = "evet",
                ["no"] = "hayır",
            },
            ["tr->en"] = new()
            {
                ["merhaba"] = "hello",
                ["güle güle"] = "goodbye",
                ["teşekkürler"] = "thank you",
                ["evet"] = "yes",
                ["hayır"] = "no",
            }
        };
    }

    /// <summary>
    /// Translate <paramref name="text"/> from <paramref name="src"/> to <paramref name="dest"/>.
    /// </summary>
    /// <exception cref="ArgumentException">Thrown when languages are unsupported or equal.</exception>
    public string Translate(string text, string src, string dest)
    {
        if (text is null) throw new ArgumentNullException(nameof(text));
        if (string.Equals(src, dest, StringComparison.OrdinalIgnoreCase))
            throw new ArgumentException("Source and destination must differ", nameof(dest));

        var key = $"{src.ToLowerInvariant()}->{dest.ToLowerInvariant()}";
        if (!_translations.ContainsKey(key))
            throw new ArgumentException($"Unsupported language pair: {src}->{dest}");

        var phrase = (text ?? string.Empty).Trim().ToLowerInvariant();
        if (_translations[key].TryGetValue(phrase, out var translated))
            return translated!; // value is non-null by construction

        // If not found, return original text unchanged
        return text;
    }
}
